rust/connlib/tunnel/src/tests/sut.rs

```diff
@@ -413,52 +380,122 @@ impl TunnelTest {
+                    continue;
+                };
+
+                on_gateway_event(*id, event, &mut self.client, now);
+                continue 'outer;
+            }
+
+            if let Some(transmit) = self.client.exec_mut(|sim| sim.sut.poll_transmit()) {
+                buffered_transmits.push_from(transmit, &self.client, now);
+                continue;
+            }
+            if let Some(event) = self.client.exec_mut(|c| c.sut.poll_event()) {
+                self.on_client_event(
+                    self.client.inner().id,
+                    event,
+                    &ref_state.portal,
+                    &ref_state.global_dns_records,
+                );
                 continue;
             }
-            break;
+            if let Some(query) = self.client.exec_mut(|client| client.sut.poll_dns_queries()) {
+                self.on_forwarded_dns_query(query, ref_state);
+                continue;
+            }
+            self.client.exec_mut(|sim| {
+                while let Some(packet) = sim.sut.poll_packets() {
+                    sim.on_received_packet(packet)
+                }
+            });
+
+            if let Some(transmit) = buffered_transmits.pop(now) {
+                self.dispatch_transmit(transmit);
+                continue;
+            }
+
+            if !buffered_transmits.is_empty() {
+                self.flux_capacitor.small_tick(); // Small tick to get to the next transmit.
+                continue;
+            }
+
+            let Some(time_to_next_action) = self.poll_timeout() else {
+                break; // Nothing to do.
+            };
+
+            if time_to_next_action > cut_off {
+                break; // Nothing to do before cut-off.
+            }
+
+            self.flux_capacitor.large_tick(); // Large tick to more quickly advance to potential next timeout.
         }
     }
 
-    /// Forwards time to the given instant iff the corresponding host would like that (i.e. returns a timestamp <= from `poll_timeout`).
-    ///
-    /// Tying the forwarding of time to the result of `poll_timeout` gives us better coverage because in production, we suspend until the value of `poll_timeout`.
-    fn handle_timeout(&mut self, now: Instant, utc_now: DateTime<Utc>) -> bool {
-        let mut any_advanced = false;
-
-        if self
-            .client
-            .exec_mut(|c| c.sut.poll_timeout())
-            .is_some_and(|t| t <= now)
-        {
-            any_advanced = true;
+    fn handle_timeout(
+        &mut self,
+        global_dns_records: &BTreeMap<DomainName, HashSet<IpAddr>>,
+        buffered_transmits: &mut BufferedTransmits,
+    ) {
+        let now = self.flux_capacitor.now();
 
-            self.client.exec_mut(|c| c.sut.handle_timeout(now));
-        };
+        while let Some(transmit) = self.client.poll_transmit(now) {
+            self.client.exec_mut(|c| c.receive(transmit, now))
+        }
```

Comment:
A `Host` now has an `inbox` in which it stores messages that have been dispatched to it but have not been fed to the actual state yet because we still need to "wait" for the inbound network latency to pass.

Once it passes, they get dispatched here.